### ISSUE TITLE
Memory leak in buffer::resize()

### ISSUE THREAD

**Issue by RElesgoe** [TheNitesWhoSay/Chkdraft#93]
Buffer.cpp

```
bool buffer::resize(s64 sizeChange, bool multiplySize)
{
    if ( this != nullptr )
    {
        if ( sizeChange > 0 && multiplySize )
            sizeAllotted = (u32)(sizeAllotted*1.2+sizeChange); // Increase allotted size by 20% + required space
        else
            sizeAllotted = (u32)(sizeAllotted+sizeChange); // Decrease allotted size by amount specified

        if ( sizeUsed > 0 )
        {
            s8* newBuffer;
            try {
                newBuffer = new s8[sizeAllotted];
            } catch ( std::bad_alloc ) {
                throw(  BadResize(0) );
            }

            if ( sizeUsed <= sizeAllotted )
                std::memcpy( newBuffer, data, sizeUsed*sizeof(char) );
            else
                std::memcpy( newBuffer, data, sizeAllotted*sizeof(char) );

            if ( data )
            {
                delete[] data;
                data = nullptr;
            }

            data = newBuffer;
        }
        else
        {
            if ( data )
            {
                delete[] data;
                data = nullptr;
            }

            if ( sizeAllotted > 0 )
            {
                try {
                    data = new s8[sizeAllotted];
                } catch ( std::bad_alloc ) {
                    sizeAllotted = 0;
                    throw( BadResize(1) );
                }
            }
        }
        return true;
    }
    return false;
}
```

Memory leak on line `data = new s8[sizeAllotted];`, memory is not freed. Use of a smart pointer is recommended.

**Comment by jjf28** [comment]
This is not a leak, it's encapsulated data taken care of in the destructor (previous memory is deleted a few lines prior), though I'll leave this open since a smart pointer would be better.

**Comment by xboi209** [comment]
I've recorded this function leaking multiple megabytes of memory over the course of less than two minutes. Will review the code in depth later.

**Comment by xboi209** [comment]
Code recommendation(unrelated to original issue):
````
if ( sizeUsed <= sizeAllotted )
	std::memcpy( newBuffer, data, sizeUsed*sizeof(char) );
else
	std::memcpy( newBuffer, data, sizeAllotted*sizeof(char) );

if ( data )
{
	delete[] data;
	data = nullptr;
}

data = newBuffer;
````

The if statement to check if variable ````data```` is not a null pointer is unnecessary. Variable ````data```` should be expected to not be a null pointer because ````std::memcpy()```` causes undefined behavior if a null pointer is used.

**Comment by xboi209** [comment]
I've attempted to rewrite Buffer.cpp using ````std::vector<s8> data;```` but the ````getPtr()```` functions seems very unsafe to have.
I've looked at the calls to ````getPtr()```` and it seems that it's unnecessary in many cases as the calling code requires a pointer to a char buffer which makes ````getString()```` more suitable. In at least one case, which is ````GRP::data()````, the function returns a raw pointer to ````data```` which doesn't seem like a good idea.
